File: claude-code-sdk/src/main/kotlin/com/claudecodeplus/sdk/protocol/ToolTypeParser.kt
package com.claudecodeplus.sdk.protocol

import com.claudecodeplus.sdk.exceptions.MessageParsingException
import com.claudecodeplus.sdk.types.*
import kotlinx.serialization.Contextual
import kotlinx.serialization.json.*

/**
 * 工具类型解析器
 *
 * 负责将通用的ToolUseBlock转换为具体的工具类型。
 * 这个解析器根据工具名称和参数创建相应的具体工具实例。
 */
object ToolTypeParser {

    /**
     * 将ToolUseBlock解析为具体的工具类型
     *
     * @param block 通用的工具使用块
     * @return 具体的工具类型实例
     */
    fun parseToolUseBlock(block: ToolUseBlock): SpecificToolUse {
        return try {
            when (block.name) {
                "Bash" -> parseBashTool(block)
                "BashOutput" -> parseBashOutputTool(block)
                "KillShell" -> parseKillShellTool(block)
                "Edit" -> parseEditTool(block)
                "MultiEdit" -> parseMultiEditTool(block)
                "Read" -> parseReadTool(block)
                "Write" -> parseWriteTool(block)
                "Glob" -> parseGlobTool(block)
                "Grep" -> parseGrepTool(block)
                "WebFetch" -> parseWebFetchTool(block)
                "WebSearch" -> parseWebSearchTool(block)
                "TodoWrite" -> parseTodoWriteTool(block)
                "Task" -> parseTaskTool(block)
                "NotebookEdit" -> parseNotebookEditTool(block)
                "ExitPlanMode" -> parseExitPlanModeTool(block)
                "ListMcpResourcesTool" -> parseListMcpResourcesTool(block)
                "ReadMcpResourceTool" -> parseReadMcpResourceTool(block)
                else -> {
                    // 检查是否是MCP工具
                    if (block.name.startsWith("mcp__")) {
                        parseMcpTool(block)
                    } else {
                        parseUnknownTool(block)
                    }
                }
            }
        } catch (e: Exception) {
            // 解析失败时回退到未知工具类型
            println("[ToolTypeParser] ⚠️ 解析工具失败 ${block.name}: ${e.message}")
            parseUnknownTool(block)
        }
    }

    /**
     * 解析Bash工具
     */
    private fun parseBashTool(block: ToolUseBlock): BashToolUse {
        val params = block.input.jsonObject
        return BashToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            command = params["command"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'command' in Bash tool"),
            description = params["description"]?.jsonPrimitive?.contentOrNull,
            timeout = params["timeout"]?.jsonPrimitive?.longOrNull,
            runInBackground = params["run_in_background"]?.jsonPrimitive?.booleanOrNull ?: false
        )
    }

    /**
     * 解析Edit工具
     */
    private fun parseEditTool(block: ToolUseBlock): EditToolUse {
        val params = block.input.jsonObject
        return EditToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            filePath = params["file_path"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'file_path' in Edit tool"),
            oldString = params["old_string"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'old_string' in Edit tool"),
            newString = params["new_string"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'new_string' in Edit tool"),
            replaceAll = params["replace_all"]?.jsonPrimitive?.booleanOrNull ?: false
        )
    }

    /**
     * 解析MultiEdit工具
     */
    private fun parseMultiEditTool(block: ToolUseBlock): MultiEditToolUse {
        val params = block.input.jsonObject
        val editsArray = params["edits"]?.jsonArray
            ?: throw MessageParsingException("Missing 'edits' in MultiEdit tool")

        val edits = editsArray.map { editElement ->
            val editObj = editElement.jsonObject
            MultiEditToolUse.EditOperation(
                oldString = editObj["old_string"]?.jsonPrimitive?.content
                    ?: throw MessageParsingException("Missing 'old_string' in edit operation"),
                newString = editObj["new_string"]?.jsonPrimitive?.content
                    ?: throw MessageParsingException("Missing 'new_string' in edit operation"),
                replaceAll = editObj["replace_all"]?.jsonPrimitive?.booleanOrNull ?: false
            )
        }

        return MultiEditToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            filePath = params["file_path"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'file_path' in MultiEdit tool"),
            edits = edits
        )
    }

    /**
     * 解析Read工具
     */
    private fun parseReadTool(block: ToolUseBlock): ReadToolUse {
        val params = block.input.jsonObject
        return ReadToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            filePath = params["file_path"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'file_path' in Read tool"),
            offset = params["offset"]?.jsonPrimitive?.intOrNull,
            limit = params["limit"]?.jsonPrimitive?.intOrNull
        )
    }

    /**
     * 解析Write工具
     */
    private fun parseWriteTool(block: ToolUseBlock): WriteToolUse {
        val params = block.input.jsonObject
        return WriteToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            filePath = params["file_path"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'file_path' in Write tool"),
            content = params["content"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'content' in Write tool")
        )
    }

    /**
     * 解析Glob工具
     */
    private fun parseGlobTool(block: ToolUseBlock): GlobToolUse {
        val params = block.input.jsonObject
        return GlobToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            pattern = params["pattern"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'pattern' in Glob tool"),
            path = params["path"]?.jsonPrimitive?.contentOrNull
        )
    }

    /**
     * 解析Grep工具
     */
    private fun parseGrepTool(block: ToolUseBlock): GrepToolUse {
        val params = block.input.jsonObject
        return GrepToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            pattern = params["pattern"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'pattern' in Grep tool"),
            path = params["path"]?.jsonPrimitive?.contentOrNull,
            outputMode = params["output_mode"]?.jsonPrimitive?.contentOrNull,
            glob = params["glob"]?.jsonPrimitive?.contentOrNull,
            type = params["type"]?.jsonPrimitive?.contentOrNull,
            caseInsensitive = params["-i"]?.jsonPrimitive?.booleanOrNull ?: false,
            showLineNumbers = params["-n"]?.jsonPrimitive?.booleanOrNull ?: false,
            contextBefore = params["-B"]?.jsonPrimitive?.intOrNull,
            contextAfter = params["-A"]?.jsonPrimitive?.intOrNull,
            headLimit = params["head_limit"]?.jsonPrimitive?.intOrNull
        )
    }

    /**
     * 解析WebFetch工具
     */
    private fun parseWebFetchTool(block: ToolUseBlock): WebFetchToolUse {
        val params = block.input.jsonObject
        return WebFetchToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            url = params["url"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'url' in WebFetch tool"),
            prompt = params["prompt"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'prompt' in WebFetch tool")
        )
    }

    /**
     * 解析WebSearch工具
     */
    private fun parseWebSearchTool(block: ToolUseBlock): WebSearchToolUse {
        val params = block.input.jsonObject
        return WebSearchToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            query = params["query"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'query' in WebSearch tool"),
            allowedDomains = params["allowed_domains"]?.jsonArray?.map { it.jsonPrimitive.content },
            blockedDomains = params["blocked_domains"]?.jsonArray?.map { it.jsonPrimitive.content }
        )
    }

    /**
     * 解析TodoWrite工具
     */
    private fun parseTodoWriteTool(block: ToolUseBlock): TodoWriteToolUse {
        val params = block.input.jsonObject
        val todosArray = params["todos"]?.jsonArray
            ?: throw MessageParsingException("Missing 'todos' in TodoWrite tool")

        val todos = todosArray.map { todoElement ->
            val todoObj = todoElement.jsonObject
            TodoWriteToolUse.TodoItem(
                content = todoObj["content"]?.jsonPrimitive?.content
                    ?: throw MessageParsingException("Missing 'content' in todo item"),
                status = todoObj["status"]?.jsonPrimitive?.content
                    ?: throw MessageParsingException("Missing 'status' in todo item"),
                activeForm = todoObj["activeForm"]?.jsonPrimitive?.content
                    ?: throw MessageParsingException("Missing 'activeForm' in todo item")
            )
        }

        return TodoWriteToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            todos = todos
        )
    }

    /**
     * 解析Task工具
     */
    private fun parseTaskTool(block: ToolUseBlock): TaskToolUse {
        val params = block.input.jsonObject
        return TaskToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            description = params["description"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'description' in Task tool"),
            prompt = params["prompt"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'prompt' in Task tool"),
            subagentType = params["subagent_type"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'subagent_type' in Task tool")
        )
    }

    /**
     * 解析NotebookEdit工具
     */
    private fun parseNotebookEditTool(block: ToolUseBlock): NotebookEditToolUse {
        val params = block.input.jsonObject
        return NotebookEditToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            notebookPath = params["notebook_path"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'notebook_path' in NotebookEdit tool"),
            newSource = params["new_source"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'new_source' in NotebookEdit tool"),
            cellId = params["cell_id"]?.jsonPrimitive?.contentOrNull,
            cellType = params["cell_type"]?.jsonPrimitive?.contentOrNull,
            editMode = params["edit_mode"]?.jsonPrimitive?.contentOrNull
        )
    }

    /**
     * 解析MCP工具
     */
    private fun parseMcpTool(block: ToolUseBlock): McpToolUse {
        // MCP工具名称格式: mcp__server_name__function_name
        val nameParts = block.name.split("__")
        val serverName = if (nameParts.size >= 2) nameParts[1] else "unknown"
        val functionName = if (nameParts.size >= 3) nameParts[2] else block.name

        // 将JsonElement转换为Map<String, Any>
        val parameters = parseJsonElementToMap(block.input)

        return McpToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            fullToolName = block.name,
            serverName = serverName,
            functionName = functionName,
            parameters = parameters
        )
    }

    /**
     * 解析BashOutput工具
     */
    private fun parseBashOutputTool(block: ToolUseBlock): BashOutputToolUse {
        val params = block.input.jsonObject
        return BashOutputToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            bashId = params["bash_id"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'bash_id' in BashOutput tool"),
            filter = params["filter"]?.jsonPrimitive?.contentOrNull
        )
    }

    /**
     * 解析KillShell工具
     */
    private fun parseKillShellTool(block: ToolUseBlock): KillShellToolUse {
        val params = block.input.jsonObject
        return KillShellToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            shellId = params["shell_id"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'shell_id' in KillShell tool")
        )
    }

    /**
     * 解析ExitPlanMode工具
     */
    private fun parseExitPlanModeTool(block: ToolUseBlock): ExitPlanModeToolUse {
        val params = block.input.jsonObject
        return ExitPlanModeToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            plan = params["plan"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'plan' in ExitPlanMode tool")
        )
    }

    /**
     * 解析ListMcpResources工具
     */
    private fun parseListMcpResourcesTool(block: ToolUseBlock): ListMcpResourcesToolUse {
        val params = block.input.jsonObject
        return ListMcpResourcesToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            server = params["server"]?.jsonPrimitive?.contentOrNull
        )
    }

    /**
     * 解析ReadMcpResource工具
     */
    private fun parseReadMcpResourceTool(block: ToolUseBlock): ReadMcpResourceToolUse {
        val params = block.input.jsonObject
        return ReadMcpResourceToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            server = params["server"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'server' in ReadMcpResource tool"),
            uri = params["uri"]?.jsonPrimitive?.content
                ?: throw MessageParsingException("Missing 'uri' in ReadMcpResource tool")
        )
    }

    /**
     * 解析未知工具
     */
    private fun parseUnknownTool(block: ToolUseBlock): UnknownToolUse {
        val parameters = parseJsonElementToMap(block.input)
        return UnknownToolUse(
            id = block.id,
            name = block.name,
            input = block.input,
            toolName = block.name,
            parameters = parameters
        )
    }

    /**
     * 将JsonElement转换为Map<String, Any>
     * 这是一个递归函数，用于处理嵌套的JSON结构
     */
    private fun parseJsonElementToMap(element: JsonElement): Map<String, @Contextual Any> {
        return when (element) {
            is JsonObject -> {
                element.mapValues { (_, value) ->
                    parseJsonValue(value)
                }
            }
            else -> mapOf("value" to parseJsonValue(element))
        }
    }

    /**
     * 递归解析JsonElement为Kotlin类型
     */
    private fun parseJsonValue(element: JsonElement): Any {
        return when (element) {
            is JsonPrimitive -> {
                when {
                    element.isString -> element.content
                    element.booleanOrNull != null -> element.boolean
                    element.intOrNull != null -> element.int
                    element.longOrNull != null -> element.long
                    element.doubleOrNull != null -> element.double
                    else -> element.content
                }
            }
            is JsonArray -> {
                element.map { parseJsonValue(it) }
            }
            is JsonObject -> {
                element.mapValues { (_, value) -> parseJsonValue(value) }
            }
            else -> element.toString()
        }
    }
}

File: vscode-extension/src/sdk/claude/protocol/toolTypeParser.ts
/**
 * Tool Type Parser for Claude CLI
 * 
 * This module parses generic ToolUseBlock into specific tool types.
 * Each tool type provides strongly-typed access to its parameters.
 */

import type { ToolUseBlock, ContentBlock } from './models';

/**
 * Tool type enumeration.
 * Defines all known Claude Code tool types.
 */
export enum ToolType {
  // Basic file operation tools
  BASH = 'CLAUDE_BASH',
  BASH_OUTPUT = 'CLAUDE_BASH_OUTPUT',
  KILL_SHELL = 'CLAUDE_KILL_SHELL',
  EDIT = 'CLAUDE_EDIT',
  MULTI_EDIT = 'CLAUDE_MULTI_EDIT',
  READ = 'CLAUDE_READ',
  WRITE = 'CLAUDE_WRITE',

  // Search and find tools
  GLOB = 'CLAUDE_GLOB',
  GREP = 'CLAUDE_GREP',

  // Network tools
  WEB_FETCH = 'CLAUDE_WEB_FETCH',
  WEB_SEARCH = 'CLAUDE_WEB_SEARCH',

  // Development and task management tools
  TODO_WRITE = 'CLAUDE_TODO_WRITE',
  TASK = 'CLAUDE_TASK',
  EXIT_PLAN_MODE = 'CLAUDE_EXIT_PLAN_MODE',
  ASK_USER_QUESTION = 'CLAUDE_ASK_USER_QUESTION',
  SKILL = 'CLAUDE_SKILL',
  SLASH_COMMAND = 'CLAUDE_SLASH_COMMAND',
  ENTER_PLAN_MODE = 'CLAUDE_ENTER_PLAN_MODE',

  // Jupyter notebook tool
  NOTEBOOK_EDIT = 'CLAUDE_NOTEBOOK_EDIT',

  // MCP (Model Context Protocol) tools
  MCP_TOOL = 'MCP',
  LIST_MCP_RESOURCES = 'CLAUDE_LIST_MCP_RESOURCES',
  READ_MCP_RESOURCE = 'CLAUDE_READ_MCP_RESOURCE',

  // Unknown tool type
  UNKNOWN = 'UNKNOWN',
}

/**
 * Get ToolType from tool name.
 */
export function getToolTypeFromName(toolName: string): ToolType {
  if (toolName.startsWith('mcp__')) {
    return ToolType.MCP_TOOL;
  }

  switch (toolName) {
    case 'Bash':
      return ToolType.BASH;
    case 'BashOutput':
      return ToolType.BASH_OUTPUT;
    case 'KillShell':
      return ToolType.KILL_SHELL;
    case 'Edit':
      return ToolType.EDIT;
    case 'MultiEdit':
      return ToolType.MULTI_EDIT;
    case 'Read':
      return ToolType.READ;
    case 'Write':
      return ToolType.WRITE;
    case 'Glob':
      return ToolType.GLOB;
    case 'Grep':
      return ToolType.GREP;
    case 'WebFetch':
      return ToolType.WEB_FETCH;
    case 'WebSearch':
      return ToolType.WEB_SEARCH;
    case 'TodoWrite':
      return ToolType.TODO_WRITE;
    case 'Task':
      return ToolType.TASK;
    case 'ExitPlanMode':
      return ToolType.EXIT_PLAN_MODE;
    case 'AskUserQuestion':
      return ToolType.ASK_USER_QUESTION;
    case 'Skill':
      return ToolType.SKILL;
    case 'NotebookEdit':
      return ToolType.NOTEBOOK_EDIT;
    case 'ListMcpResourcesTool':
      return ToolType.LIST_MCP_RESOURCES;
    case 'ReadMcpResourceTool':
      return ToolType.READ_MCP_RESOURCE;
    default:
      return ToolType.UNKNOWN;
  }
}

// ============================================================================
// Specific Tool Use Interfaces
// ============================================================================

/**
 * Base interface for specific tool uses.
 */
export interface SpecificToolUse extends ContentBlock {
  type: 'tool_use';
  id: string;
  name: string;
  input: unknown;
  toolType: ToolType;
}

/**
 * Bash tool use - execute shell commands.
 */
export interface BashToolUse extends SpecificToolUse {
  command: string;
  description?: string;
  timeout?: number;
  runInBackground?: boolean;
}

/**
 * BashOutput tool use - get output from a Bash command.
 */
export interface BashOutputToolUse extends SpecificToolUse {
  bashId: string;
  filter?: string;
}

/**
 * KillShell tool use - terminate a running shell process.
 */
export interface KillShellToolUse extends SpecificToolUse {
  shellId: string;
}

/**
 * Edit tool use - edit file content.
 */
export interface EditToolUse extends SpecificToolUse {
  filePath: string;
  oldString: string;
  newString: string;
  replaceAll?: boolean;
}

/**
 * Edit operation for MultiEdit.
 */
export interface EditOperation {
  oldString: string;
  newString: string;
  replaceAll?: boolean;
}

/**
 * MultiEdit tool use - multiple edits to a single file.
 */
export interface MultiEditToolUse extends SpecificToolUse {
  filePath: string;
  edits: EditOperation[];
}

/**
 * Read tool use - read file content.
 */
export interface ReadToolUse extends SpecificToolUse {
  filePath: string;
  offset?: number;
  limit?: number;
}

/**
 * Write tool use - write file content.
 */
export interface WriteToolUse extends SpecificToolUse {
  filePath: string;
  content: string;
}

/**
 * Glob tool use - file pattern matching.
 */
export interface GlobToolUse extends SpecificToolUse {
  pattern: string;
  path?: string;
}

/**
 * Grep tool use - text search.
 */
export interface GrepToolUse extends SpecificToolUse {
  pattern: string;
  path?: string;
  outputMode?: string;
  glob?: string;
  grepType?: string;
  caseInsensitive?: boolean;
  showLineNumbers?: boolean;
  contextBefore?: number;
  contextAfter?: number;
  headLimit?: number;
}

/**
 * WebFetch tool use - fetch web content.
 */
export interface WebFetchToolUse extends SpecificToolUse {
  url: string;
  prompt: string;
}

/**
 * WebSearch tool use - web search.
 */
export interface WebSearchToolUse extends SpecificToolUse {
  query: string;
  allowedDomains?: string[];
  blockedDomains?: string[];
}

/**
 * Todo item for TodoWrite.
 */
export interface TodoItem {
  content: string;
  status: string; // 'pending' | 'in_progress' | 'completed'
  activeForm: string;
}

/**
 * TodoWrite tool use - manage todo list.
 */
export interface TodoWriteToolUse extends SpecificToolUse {
  todos: TodoItem[];
}

/**
 * Task tool use - launch sub-task agent.
 */
export interface TaskToolUse extends SpecificToolUse {
  description: string;
  prompt: string;
  subagentType: string;
}

/**
 * NotebookEdit tool use - edit Jupyter Notebook.
 */
export interface NotebookEditToolUse extends SpecificToolUse {
  notebookPath: string;
  newSource: string;
  cellId?: string;
  cellType?: string;
  editMode?: string;
}

/**
 * ExitPlanMode tool use - exit plan mode.
 */
export interface ExitPlanModeToolUse extends SpecificToolUse {
  plan: string;
}

/**
 * MCP tool use - Model Context Protocol tool.
 */
export interface McpToolUse extends SpecificToolUse {
  fullToolName: string;
  serverName: string;
  functionName: string;
  parameters: Record<string, unknown>;
}

/**
 * ListMcpResources tool use - list MCP server resources.
 */
export interface ListMcpResourcesToolUse extends SpecificToolUse {
  server?: string;
}

/**
 * ReadMcpResource tool use - read MCP resource.
 */
export interface ReadMcpResourceToolUse extends SpecificToolUse {
  server: string;
  uri: string;
}

/**
 * Skill tool use - invoke registered skills.
 */
export interface SkillToolUse extends SpecificToolUse {
  skill: string;
  args?: string;
}

/**
 * Unknown tool use - for unrecognized tool types.
 */
export interface UnknownToolUse extends SpecificToolUse {
  toolName: string;
  parameters: Record<string, unknown>;
}

// ============================================================================
// Tool Type Parser
// ============================================================================

/**
 * Tool Type Parser - converts generic ToolUseBlock to specific tool types.
 */
export const ToolTypeParser = {
  /**
   * Parse ToolUseBlock into a specific tool type.
   */
  parseToolUseBlock(block: ToolUseBlock): SpecificToolUse {
    try {
      switch (block.name) {
        case 'Bash':
          return this.parseBashTool(block);
        case 'BashOutput':
          return this.parseBashOutputTool(block);
        case 'KillShell':
          return this.parseKillShellTool(block);
        case 'Edit':
          return this.parseEditTool(block);
        case 'MultiEdit':
          return this.parseMultiEditTool(block);
        case 'Read':
          return this.parseReadTool(block);
        case 'Write':
          return this.parseWriteTool(block);
        case 'Glob':
          return this.parseGlobTool(block);
        case 'Grep':
          return this.parseGrepTool(block);
        case 'WebFetch':
          return this.parseWebFetchTool(block);
        case 'WebSearch':
          return this.parseWebSearchTool(block);
        case 'TodoWrite':
          return this.parseTodoWriteTool(block);
        case 'Task':
          return this.parseTaskTool(block);
        case 'NotebookEdit':
          return this.parseNotebookEditTool(block);
        case 'ExitPlanMode':
          return this.parseExitPlanModeTool(block);
        case 'ListMcpResourcesTool':
          return this.parseListMcpResourcesTool(block);
        case 'ReadMcpResourceTool':
          return this.parseReadMcpResourceTool(block);
        case 'Skill':
          return this.parseSkillTool(block);
        default:
          if (block.name.startsWith('mcp__')) {
            return this.parseMcpTool(block);
          }
          return this.parseUnknownTool(block);
      }
    } catch (e) {
      console.warn(`[ToolTypeParser] Failed to parse tool ${block.name}:`, e);
      return this.parseUnknownTool(block);
    }
  },

  /**
   * Get input as object.
   */
  getInputObject(input: unknown): Record<string, unknown> {
    if (typeof input === 'object' && input !== null) {
      return input as Record<string, unknown>;
    }
    return {};
  },

  parseBashTool(block: ToolUseBlock): BashToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.BASH,
      command: (params.command as string) || '',
      description: params.description as string | undefined,
      timeout: params.timeout as number | undefined,
      runInBackground: (params.run_in_background as boolean) ?? false,
    };
  },

  parseBashOutputTool(block: ToolUseBlock): BashOutputToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.BASH_OUTPUT,
      bashId: (params.bash_id as string) || '',
      filter: params.filter as string | undefined,
    };
  },

  parseKillShellTool(block: ToolUseBlock): KillShellToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.KILL_SHELL,
      shellId: (params.shell_id as string) || '',
    };
  },

  parseEditTool(block: ToolUseBlock): EditToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.EDIT,
      filePath: (params.file_path as string) || '',
      oldString: (params.old_string as string) || '',
      newString: (params.new_string as string) || '',
      replaceAll: (params.replace_all as boolean) ?? false,
    };
  },

  parseMultiEditTool(block: ToolUseBlock): MultiEditToolUse {
    const params = this.getInputObject(block.input);
    const editsArray = params.edits as unknown[] | undefined;

    const edits: EditOperation[] = (editsArray ?? []).map((edit) => {
      const editObj = edit as Record<string, unknown>;
      return {
        oldString: (editObj.old_string as string) || '',
        newString: (editObj.new_string as string) || '',
        replaceAll: (editObj.replace_all as boolean) ?? false,
      };
    });

    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.MULTI_EDIT,
      filePath: (params.file_path as string) || '',
      edits,
    };
  },

  parseReadTool(block: ToolUseBlock): ReadToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.READ,
      filePath: (params.file_path as string) || '',
      offset: params.offset as number | undefined,
      limit: params.limit as number | undefined,
    };
  },

  parseWriteTool(block: ToolUseBlock): WriteToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.WRITE,
      filePath: (params.file_path as string) || '',
      content: (params.content as string) || '',
    };
  },

  parseGlobTool(block: ToolUseBlock): GlobToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.GLOB,
      pattern: (params.pattern as string) || '',
      path: params.path as string | undefined,
    };
  },

  parseGrepTool(block: ToolUseBlock): GrepToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.GREP,
      pattern: (params.pattern as string) || '',
      path: params.path as string | undefined,
      outputMode: params.output_mode as string | undefined,
      glob: params.glob as string | undefined,
      grepType: params.type as string | undefined,
      caseInsensitive: (params['-i'] as boolean) ?? false,
      showLineNumbers: (params['-n'] as boolean) ?? false,
      contextBefore: params['-B'] as number | undefined,
      contextAfter: params['-A'] as number | undefined,
      headLimit: params.head_limit as number | undefined,
    };
  },

  parseWebFetchTool(block: ToolUseBlock): WebFetchToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.WEB_FETCH,
      url: (params.url as string) || '',
      prompt: (params.prompt as string) || '',
    };
  },

  parseWebSearchTool(block: ToolUseBlock): WebSearchToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.WEB_SEARCH,
      query: (params.query as string) || '',
      allowedDomains: params.allowed_domains as string[] | undefined,
      blockedDomains: params.blocked_domains as string[] | undefined,
    };
  },

  parseTodoWriteTool(block: ToolUseBlock): TodoWriteToolUse {
    const params = this.getInputObject(block.input);
    const todosArray = params.todos as unknown[] | undefined;

    const todos: TodoItem[] = (todosArray ?? []).map((todo) => {
      const todoObj = todo as Record<string, unknown>;
      return {
        content: (todoObj.content as string) || '',
        status: (todoObj.status as string) || 'pending',
        activeForm: (todoObj.activeForm as string) || '',
      };
    });

    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.TODO_WRITE,
      todos,
    };
  },

  parseTaskTool(block: ToolUseBlock): TaskToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.TASK,
      description: (params.description as string) || '',
      prompt: (params.prompt as string) || '',
      subagentType: (params.subagent_type as string) || '',
    };
  },

  parseNotebookEditTool(block: ToolUseBlock): NotebookEditToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.NOTEBOOK_EDIT,
      notebookPath: (params.notebook_path as string) || '',
      newSource: (params.new_source as string) || '',
      cellId: params.cell_id as string | undefined,
      cellType: params.cell_type as string | undefined,
      editMode: params.edit_mode as string | undefined,
    };
  },

  parseExitPlanModeTool(block: ToolUseBlock): ExitPlanModeToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.EXIT_PLAN_MODE,
      plan: (params.plan as string) || '',
    };
  },

  parseListMcpResourcesTool(block: ToolUseBlock): ListMcpResourcesToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.LIST_MCP_RESOURCES,
      server: params.server as string | undefined,
    };
  },

  parseReadMcpResourceTool(block: ToolUseBlock): ReadMcpResourceToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.READ_MCP_RESOURCE,
      server: (params.server as string) || '',
      uri: (params.uri as string) || '',
    };
  },

  parseSkillTool(block: ToolUseBlock): SkillToolUse {
    const params = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.SKILL,
      skill: (params.skill as string) || '',
      args: params.args as string | undefined,
    };
  },

  parseMcpTool(block: ToolUseBlock): McpToolUse {
    // MCP tool name format: mcp__server_name__function_name
    const nameParts = block.name.split('__');
    const serverName = nameParts.length >= 2 ? nameParts[1] : 'unknown';
    const functionName = nameParts.length >= 3 ? nameParts[2] : block.name;

    const parameters = this.getInputObject(block.input);

    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.MCP_TOOL,
      fullToolName: block.name,
      serverName,
      functionName,
      parameters,
    };
  },

  parseUnknownTool(block: ToolUseBlock): UnknownToolUse {
    const parameters = this.getInputObject(block.input);
    return {
      type: 'tool_use',
      id: block.id,
      name: block.name,
      input: block.input,
      toolType: ToolType.UNKNOWN,
      toolName: block.name,
      parameters,
    };
  },
};
